soql-analyzer/src/main/scala/com/socrata/soql/analyzer2/rewrite/PreserveOrdering.scala

```diff
@@ -0,0 +1,124 @@
+  // thing, it will not necessarily get it!
+  def rewriteStatement(stmt: Statement, wantOutputOrdered: Boolean, wantOrderingColumns: Boolean): (Seq[(ColumnLabel, CT, Boolean, Boolean)], Statement) = {
+    stmt match {
+      case CombinedTables(op, left, right) =>
+        // table ops never preserve ordering
+        (Nil, CombinedTables(op, rewriteStatement(left, false, false)._2, rewriteStatement(right, false, false)._2))
+
+      case cte@CTE(defLabel, defAlias, defQuery, matHint, useQuery) =>
+        val (orderingColumns, newUseQuery) = rewriteStatement(useQuery, wantOutputOrdered, wantOrderingColumns)
+
+        (
+          orderingColumns,
+          cte.copy(
+            definitionQuery = rewriteStatement(defQuery, true, false)._2,
+            useQuery = newUseQuery
+          )
+        )
+
+      case v@Values(_) =>
+        // TODO, should this rewrite into a SELECT ... FROM
+        // (values...) ORDER BY synthetic_column ?  If so, we'll need
+        // a way to generate synthetic_column (== have a way to turn
+        // an integer into a CV)
+        (Nil, v)
+
+      case select@Select(distinctiveness, selectList, from, where, groupBy, having, orderBy, limit, offset, search, hint) =>
+        val usedNames = selectList.valuesIterator.map(_.name).to(scm.HashSet)
+        def freshName(base: String) = {
+          val created = Iterator.from(1).map { i => ColumnName(base + "_" + i) }.find { n =>
+            !usedNames.contains(n)
+          }.get
+          usedNames += created
+          created
+        }
+
+        // If we're windowed, we want the underlying query ordered if
+        // possible even if our caller doesn't care, unless there's an
+        // aggregate in the way, in which case the aggregate will
+        // destroy any underlying ordering anyway so we stop caring.
+
+        val wantSubqueryOrdered = (select.isWindowed || wantOutputOrdered) && !select.isAggregated && distinctiveness == Distinctiveness.Indistinct
+        val (extraOrdering, newFrom) = rewriteFrom(from, wantSubqueryOrdered, wantSubqueryOrdered)
+
+        // We will at the very least want to add the ordering
+        // columns from the subquery onto the end of our order-by...
+        val orderedSelf = select.copy(
+          from = newFrom,
+          orderBy = orderBy ++ extraOrdering
+        )
+        if(wantOrderingColumns) {
+          // ..and our caller wants to know how we're ordered, so
+          // make sure we've put the relevant expressins in our
+          // select list and return them.
+          val (newColumns, outputInfo) = orderedSelf.orderBy.map { case OrderBy(expr, asc, nullLast) =>
+            val columnLabel = provider.columnLabel()
+            (columnLabel -> NamedExpr(expr, freshName("order")), (columnLabel, expr.typ, asc, nullLast))
+          }.unzip
+
+          val newSelf = orderedSelf.copy(selectList = selectList ++ newColumns)
+
+          (outputInfo, newSelf)
```

Comment:
This currently just creates new output columns to reflect the ordering; it could in principle check to see if those exprs are already being selected and just re-use them.  Perhaps worth it...

Reply:
yes, I decided it was worth it.